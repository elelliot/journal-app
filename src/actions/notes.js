import { db } from "../firebase/firebase-config";
import { collection, addDoc, updateDoc, doc, deleteDoc } from "firebase/firestore";

import { types } from '../types/types';
import { loadNotes } from "../helpers/loadNotes";
import Swal from 'sweetalert2';
import { fileUpload } from "../helpers/fileUpload";

//Estructura de la db de las notas: uid (user) -> journal -> notes -> id( autogenerado) de la nota -> nota


export const startNewNote = () => {

    return async ( dispatch, getState ) => {

        //obtenemos el State, literal todo, y con esto podemos obtener cualquier cosa del estado
        // const state = getState();
        // console.log(state);

        //obtenemos uid del user (obtenido desde el state) 
        const { uid } = getState().auth;
        
        //nueva nota
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime() //momento exacto en que se crea la nota
        }

        //Query, no es obligatorio, pero es buena práctica hacerlo con try catch
        //agregamos a la db, la nueva nota con id autogenerado en la colección dada (si no existe, la crea).

        let docRef = '';
        try {
            docRef = await addDoc(collection(db, `${uid}/journal/notes`), newNote)
            //console.log("New Note Added",docRef);
          } catch (e) {
            console.log("Error: Couldn't create Note",e)
          }


        //mandamos al reducer
        dispatch( activeNote( docRef.id, newNote ) );

        //Refresh Sidebar
        dispatch( addNewNote(docRef.id, newNote) );
    }
}

//ya creamos la nota, ahora la queremos visualizar al crearla
export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id, //id de la nota
        ...note //el resto igual con todo y id
    }
});

//para que al crear la nota, haga refresh el sidebar
export const addNewNote = ( id,note ) => ({
    type: types.notesAddNew,
    payload: {
        id, ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async( dispatch ) => {
        //loadNotes es una promesa..., y después 
        const notes = await loadNotes( uid );
        //aquí es cuando se mandan a Redux
        dispatch( setNotes( notes ) );
    }
}
export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes //como siempre estamos creando un nuevo arreglo, no necesitamos el spread ya que sería "redundante"
});


//Update en la BD al dar click en Save
export const startSaveNote = ( note ) => {
    return async( dispatch, getState ) => {

        //obtenemos uid del user (obtenido desde el state) 
        const { uid } = getState().auth;

        //Si no viene el url de la imagen, lo borramos para evitar meter un undefined a firebase ya que dará error
        if( !note.url ) {
            delete note.url;
        }


        //El id de la nota no lo queremos grabar en FireStore, por eso lo eliminamos en nu nuevo obj.
        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        //Update
        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        try {
            await updateDoc(noteRef, noteToFireStore);
            //console.log('Note Updated!');
            
            //de esta forma el sideBar se actualiza al oprimir save, el problema es que además de la nota que cambió, también se actualizan las demás
            //dispatch(startLoadingNotes( uid );
            dispatch( refreshNote( note.id, noteToFireStore) );

            //Saved Note Alert
            Swal.fire({
                iconColor: '#41D70C',
                icon: 'success',
                titleText: `Note: "${ noteToFireStore.title }" Saved!`,
                color: "#FFFFFF",
                background: "#4D4D4D",
              })


          } catch (e) {
            console.log('There was an error while updating note!',e);
          }
    }

}

//
export const refreshNote = ( id, note ) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note //Recordar que a noteToFirestore le borramos el id  
        }
    }
})


//subimos la imagen (es asíncrona)
export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        
        //obtenemos la activeNote y la renombramos
        const { active: activeNote } = getState().notes;

        //Para que se muestre "cargando" al subir imagen...
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });
        
        //Subimos el file a Cloudinary y obtenemos el url de la imagen subida a Cloudinary
        const fileUrl = await fileUpload( file );
        //console.log(fileUrl);

        //Le asignamos el url de la imagen a la nota activa y lo mandamos con el dispatch a FireStore
        activeNote.url = fileUrl;
        await dispatch( startSaveNote( activeNote ) );

        Swal.close();//quitamos la carga en cuanto la imagen se haya subido a Cloudinary
    }
}


//Borrar la nota de FireStore
export const startDeleting = ( id ) => {

    return async( dispatch, getState ) => {

        //obtenemos uid de la nota actual
        const { uid } = getState().auth;

        //referencia a la nota activa(id) del usuario (uid)
        const noteRef = doc(db, `${uid}/journal/notes/${id}`);
        await deleteDoc(noteRef); //borramos la nota
        
        //Modificamos el Store
        dispatch( deleteNote(id) );
    }
}

//Borramos la nota del Store
export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id
});


//Limpiar las notas del state al dar logout
export const noteLogout = () => ({
    type: types.notesLogoutCleaning
});