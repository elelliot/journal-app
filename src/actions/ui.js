import { Toast } from "../alerts/alert";
import { types } from "../types/types";

//No necesitamos que sean async actions
// export const setError = ( err ) => ({
//     type: types.uiSetError,
//     payload: err
// });

export const setError = (err) =>{
    Toast.fire({
        icon: 'error',
        title: err
      });
      return{
        type: types.uiSetError,
        payload: err
      }
}


export const removeError = () => ({
    type: types.uiRemoveError
});

//Controlamos el estado del loading mandando el type de accion
export const startLoading = () => ({
    type: types.uiStartLoading
});


export const finishLoading = () => ({
    type: types.uiFinishLoading
});