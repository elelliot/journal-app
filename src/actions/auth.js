

import { getAuth, signInWithPopup, createUserWithEmailAndPassword, signInWithEmailAndPassword, updateProfile, signOut } from "firebase/auth";
import { Toast } from "../alerts/alert";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

//Alerts ver en alert.js
const welcome = () => { Toast.fire({
    iconColor: '#41D70C',
    icon: 'success',
    title: 'Welcome, what do we have for today? ^u^/'
  })};

const loginErr = () => {Toast.fire({
    icon: 'error',
    title: 'User or Password might be wrong or don\'t exist, try again ^u^'
  })};


const emailUsed = () => {Toast.fire({
    icon: 'error',
    title: 'Email already used'
  })};
//actions


//acción que dispara otra acción(login) cuando el setTimeout(tarea asíncrona) se resuelve
export const startLoginEmailPassword = (email, password) => {

    //al usar el dispatch de redux, la funcion retorna un callback ()=> ,y uno de sus args es el dispatch de Thunk  (todo adentro del callback se ejecuta)
    //se espera que se use el dispatch para trabajar los datos que llegan de una tarea async

    //se pueden realizar fetch, posteo de archivos o incluso dispatch de otra acción dentro del callback
    return (dispatch) => {
    
        //loading: true
        dispatch( startLoading() );

        //Para loguearse con email y password
        const auth = getAuth();
           return signInWithEmailAndPassword(auth, email, password)
            .then( ({user}) => {
                //console.log(user)
                //Mandamos uid y displayName al state cuando se hace login
                //dispatch( login( user.uid, user.displayName) );
                
                //PA LAS PRUEBAS puse el return y el dispatch del login(con esto se dispararía 2 veces en la app):
                dispatch( login( user.uid, user.displayName) );
                
                welcome()
            })
            .catch( ({ message }) => {
                
                //Args: Title, HTML, Icon... Alerta de error de Sweet Alert2
                //Swal.fire('Error', message, 'error');
                
                loginErr()
            })
            .finally( () => {
                //Para no tener que usar el finishLoading 2 veces podemos usarlo una vez aquí y en cualquiera de los 2 casos se llama
                //loading: false
                dispatch( finishLoading() );
            });
    }
}



//Registrar email
export const startRegisterWithEmail = ( email, password, name ) => {
    //en cuanto tenga el user grabado en firebase realizaré el dispatch
    return( dispatch ) => {

        const auth = getAuth();
        //en cuanto se crea el user esto intenta autenticar inmediatamente por eso da error por ahora.
        createUserWithEmailAndPassword(auth, email, password)
            .then( async ({ user }) => {
                
                //"displayName" es null usando esta función, solo se obtiene cuando se usa una red social como login
                //podemos asignar el name(el que damos en el form) al displayName con updateProfile
                await updateProfile(user,{displayName: name})
                
                //ya que el useEffect se ejecuta antes del updateProfile, el name es null, así que por ahora haremos el mismo dispatch con el nombre obtenido del update y tendremos 2 login al terminar el registro
                dispatch( login(user.uid, user.displayName ) )
                welcome()
            })
            .catch( ({ code } ) => {
                //desestructuramos message de error
                
                emailUsed()
            });
    }
}



//login con google 
export const startGoogleLogin = () =>{
    return (dispatch) =>{
        const auth = getAuth();
        //(aquí me tira el popup de login)
        signInWithPopup(auth, googleAuthProvider)
        //se obtiene el parametro por default y desestructuramos user del mismo
            .then( ({ user }) =>{
                //desestructuramos para obtener uid y displayName (user es solo una propiedad del objeto grande que te da la promise al cumplirse, por eso se ponen así en específico):
                //dispatch( login( user.uid, user.displayName ) );
                
                welcome()
            })
            .catch( ( { code } )  => {
                //por si se cierra el popup sin haber seleccionado usuario
                
                // const errorCode = error.code;
                // const errorMessage = error.message;
                //console.log(message);
                Toast.fire({
                    icon: 'error',
                    title: code
                  })
            });
    }
}


//El login que se usa en las otras 2 acciones
export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid, 
        displayName
    }
});


//Logout es Async por que hay que se usa una función firebase para esto y usa promesa

export const startLogout = () => {
    return async( dispatch ) => {
        const auth = getAuth();
        await signOut(auth)
            .then( ()=>{

                //limpiamos el user del state 
                dispatch( logout() );
                //limpiamos las notas del state
                dispatch( noteLogout() );
            })
            .catch( (e) => {
                console.log(e); //no suele pasar pero en caso de que se vaya el internet o algo
            })
    }
}

export const logout = () => ({
    type: types.logout,
});


//forma larga del return
// export const login = (uid, displayName) => {
//     return {
//         type: types.login,
//         payload: {
//             uid, 
//             displayName
//         }
//     }
// }


