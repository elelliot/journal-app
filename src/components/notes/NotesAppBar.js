import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active } = useSelector( state => state.notes );

    //Mandamos la nota Activa para grabarla/actualizarla
    const handleSave = () => {
        //console.log(active);
        dispatch( startSaveNote( active ) )
    }

    //Para subir imagen, con esto hacemos que el botón trabaje como input file y no usarlo tal cual por razones de diseño etc.
    const handlePictureUpload = () => {
        //seleccionamos el input file al dar click al boton de picture y simulamos el click ahí, por eso es invisible.
        document.querySelector('#fileSelector').click();

    }

    const handleFileChange = (e) => {
        //console.log(e.target.files); //pa ver el archivo que enviamos,también puede ser length 0 si no enviamos la imagen
        const file = e.target.files[0]; //tendrá indice 0 si se subió la imagen

        if ( file ) {
            //mandamos el archivo a la action
            dispatch( startUploading( file ) );
        }
    }

    return (
        <div className="notes_appbar">
            <span>28 de Agosto 2021</span>

            <input 
                id = "fileSelector"
                type="file"
                name="file"
                style={ { display: 'none' } }
                onChange={ handleFileChange }
            />
            
            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureUpload }
                    >
                    Picture
                </button>
                <button 
                    className="btn"
                    onClick={ handleSave }
                >
                    Save
                </button>
            </div>
        </div>
    )
}


/*Para subir imagenes con Cloudinary: https://api.cloudinary.com/v1_1/dnumevud1/upload , ese es nuestro url (Ver documentación)
y es POST (Pruebas con POSTMAN)
Necesitamos 2keys y values que agregaremos en Body/form-data:
Keys: file (Lo seteamos en tipo file) ----- upload_preset (tipo texto)
Values: Seleccionamos la imagen que queramos ----- react-journal (nombre del upload preset que usaremos)

-Obtenemos la imagen subida del secureUrl que te da el resultado 
*/