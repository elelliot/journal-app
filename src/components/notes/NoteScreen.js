import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
    const dispatch = useDispatch();

    //referencia a la nota activa, renombramos active a note
    const { active:note } = useSelector( state => state.notes );
    
    //Para poder mostrar la info en los inputs y no tener que modificar directamente la nota activa
    /*Solo el store cambia en cada click a una nueva nota, pero este state no cambia por si mismo y se queda con los mismos valores, 
    para eso es el useEffect, para resetear el form con los nuevos valores de la nota*/
    const [ formValues, handleInputChange,reset ]= useForm( note );
    
    const { body, title, id } = formValues;

    //id de la nota Seleccionada actualmente
    const activeId = useRef( note.id )//almacenamos una variable mutable que no redibuja todo el comp si cambia
    
    /*se dispara solo si el note.id cambió (sería al darle click) y reseteamos el form con los nuevos valores, 
    para eso hicimos el cambio al reset del useForm*/ 
    useEffect(() => {

        if( note.id !== activeId.current ) {
            reset( note );
            //establecemos la nueva nota activa si no, la condicion siempre ejecutaria el effect
            activeId.current = note.id
        }

    }, [reset, note]);

    //Mandamos el cambio del form a la acción y la tendremos en Redux
    //Esto me ejecuta la acción 2 veces al seleccionar la nota
    useEffect(() => {
      
        //console.log(formValues);//afuera se dispararía con el state anterior y el nuevo por el effect de arriba
        dispatch( activeNote( formValues.id, { ...formValues } ) )//no importa que le mandemos el id con el spread a la accion, se sobreescribe en la misma, así que no problem 
    
    }, [formValues, dispatch]);
    
    
    
    //Borrar nota
    const handleDelete = () => {
        
        dispatch( startDeleting(id) );
    }
    
    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">

                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value= { title }
                    onChange={ handleInputChange }
                />

                <textarea
                    placeholder="What happened today"
                    className="notes__textarea"
                    name="body"
                    value= { body }
                    onChange={ handleInputChange }
                >

                </textarea>

                {
                    //Mostramos la imagen si existe
                    (note.url) &&
                    <div className="notes__image">
                        <img 
                            src={ note.url }
                            alt="imagen"
                        />
                    </div>
                }
                
            </div>

            <button
                className='btn btn-danger'
                onClick={ handleDelete }   
            >
                Delete
            </button>
        </div>
    )
}
