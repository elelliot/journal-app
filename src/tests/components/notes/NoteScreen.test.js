import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { activeNote } from "../../../actions/notes";
import { NoteScreen } from "../../../components/notes/NoteScreen";

jest.mock('../../../actions/notes', ()=> ({
    activeNote: jest.fn()
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

//No tiene que ser legit la info, pero con que simule el state que necesitamos
const initState = {
    auth: {
        uid:'1232',
        name: 'ElElliot'
    },
    ui:{
        loading: false,
        msgError: null
    },
    notes:{
        notes:[],
        active:{
            id:'1234ggg',
            body:'Friend of a Friend',
            date:0,
            title:'Hello'
        }
    }
};



let store = mockStore( initState );
store.dispatch = jest.fn();


const wrapper = mount(
    <Provider store={ store }>
        <NoteScreen />
    </Provider>
);


describe('Pruebas en <NoteScreen />', () => {
    
    test('Debe mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe de disparar el activeNote ', () => { 
        //Simulamos el cambio al title de la activeNote
        wrapper.find('input[name="title"]').simulate('change',{
            target:{
                name:'title',
                value:'Hello, Again'
            }
        });

        //expect(activeNote).toHaveBeenCalledTimes(1);

        //cuando se llega a llamar varias veces, en este caso solo importa la última llamada, pero como se llamó solo una vez no importa.
        // expect( activeNote ).toHaveBeenLastCalledWith( 
        //     '1234ggg',
        //     {
        //         body:'Friend of a Friend',
        //         title:'Hello, Again',
        //         id:'1234ggg',
        //         date: 0

        //     }
        // );
        
        expect( activeNote ).toHaveBeenCalledWith( 
            '1234ggg',
            {
                body:'Friend of a Friend',
                title:'Hello, Again',
                id:'1234ggg',
                date: 0

            }
        );


     })


})