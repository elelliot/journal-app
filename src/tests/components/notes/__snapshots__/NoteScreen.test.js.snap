// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> Debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes_appbar"
  >
    <span>
      28 de Agosto 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hello"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Friend of a Friend"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
